### PR TITLE
fix: update visibility check with overflow: hidden and element on the edge

### LINKED ISSUE #27514: Visibility check not working for close elements
### Current behavior

The problem is best described with this simple example - it is a component-testing of [ngx-gallery](https://github.com/daelmaak/ngx-gallery):

Screenshot 2023-08-10 at 11 00 58
 
The gallery here displays one image at a time. In the test, I slide from the 1st image to the 2nd one and check that the first one is `not.be.visible` after.
 
Cypress in this case checks the coordinates of the element in question and checks whether the [element is outside of its parent position](https://github.com/cypress-io/cypress/blob/b17043102bcb7bb653efb6526238468c74b4d564/packages/driver/src/dom/visibility.ts#L330) and if the parent also has `overflow: hidden`, it can conclude that the element in question indeed isn’t visible. Which should be my case exactly because one of the parents (`viewer` element) thinly wraps the current image and completely hides the others with the overflow: hidden.
 
But cypress thinks it is visible because of the `<` from the line I shared in the link is not sufficient. It has to be `<=` in order to work correctly. In my case, the rightmost x-position of the 1st image is compared to the leftmost x-position of the parent, and both have the value of 50px. This looks like the images overlap by 1px, but they don’t, they are placed neatly next to each other. So due to cypress’s implementation of finding the x-positions, `<=` needs to be used instead of `<`. I am happy to implement this fix if you'd like.
 

### Desired behavior

Cypress recognizes that the 2 elements - images - are only "touching" but not overlapping and the 1st is therefore not visible.

### Test code to reproduce

https://github.com/daelmaak/ngx-gallery/blob/test/introduce-cypress-for-component-testing/libs/gallery/src/lib/components/viewer/viewer.component.cy.ts

### Cypress Version

12.16

### Node version

20.5.0

### Operating System

macOS 13.5

### Debug Logs

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by canourrea23**
- Closes #27514

### Additional details

Assuming the following html:
```html
<div style='width: 100px; height: 100px; overflow: hidden; position: relative;'>
  <span style='position: absolute; width: 100px; height: 100px; left: -100px; top: 0;'>position: absolute, out of bounds left</span>
</div>
```
Previously the span would return as visible when it is butted up right against the parent but still outside of it. Updated the visibility check to handle the scenario when the elements are butted up against each other.

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by nagash77** [comment]
@canourrea23 can you please accept the CLA?

**Comment by nagash77** [comment]
You will also need to fix your commit message to include the fix: prefix to follow semantic versioning.  I would suggest reading over our [Contributing Doc](https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#pull-requests) for details.

**Comment by daelmaak** [comment]
This has been pending for 2 weeks for no good reason, can somebody please take over and get it over the finish line?

**Review by jennifer-shehane — DISMISSED** [review]
@canourrea23 Thanks! We'll need a test around the new behavior before we accept this fix.

There is a Contributing Guide that covers how to contribute and get Cypress running locally in generally here: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md

Here's an example of another issue's test code:

- HTML template replicating bug: https://github.com/cypress-io/cypress/blob/develop/packages/driver/cypress/fixtures/issue-565.html
- Test spec: https://github.com/cypress-io/cypress/blob/develop/packages/driver/cypress/e2e/issues/565.cy.js

Instructions for running the `driver` tests can always be found here: https://github.com/cypress-io/cypress/blob/develop/packages/driver/README.md

**Comment by canourrea23** [comment]
Hey @jennifer-shehane it looks like the first 3 links aren't opening for me.

**Comment by jennifer-shehane** [comment]
@canourrea23 Sorry, fixed the links.

File: packages/driver/src/dom/visibility.ts [hunk 1]
```diff
@@ -552,8 +555,13 @@ export const getReasonIsHidden = function ($el, options = { checkOpacity: true }
 
   return `This element \`${node}\` is not visible.`
 }
-/* eslint-enable no-cond-assign */
```
Comment (chrisbreiding) [inline packages/driver/src/dom/visibility.ts:555]: Is there a reason for removing this?

File: packages/driver/cypress/e2e/dom/visibility.cy.ts [hunk 2]
```diff
@@ -819,10 +819,9 @@ describe('src/cypress/dom/visibility', () => {
     })
 
     describe('css overflow', () => {
-      it('is visible when parent doesnt have overflow hidden', function () {
-        expect(this.$parentNoWidthHeightOverflowAuto.find('span')).to.be.visible
-
-        expect(this.$parentNoWidthHeightOverflowAuto.find('span')).to.not.be.hidden
+      it('is hidden when parent overflow auto and no width/height', function () {
```
Comment (chrisbreiding) [inline packages/driver/cypress/e2e/dom/visibility.cy.ts:822]: Is this removed test or some form of it no longer valid?

File: cli/CHANGELOG.md [hunk 3]
```diff
@@ -1,4 +1,12 @@
 <!-- See the ../guides/writing-the-cypress-changelog.md for details on writing the changelog. -->
+## 13.3.4
+
+_Released 11/7/2023 (PENDING)_
+
+**Bugfixes:**
+
+- Fixed an issue where an element's visibility would return visible when it was actually hidden. Fixes [#27514](https://github.com/cypress-io/cypress/issues/27514).
```
Comment (chrisbreiding) [inline cli/CHANGELOG.md]:
```suggestion
- Fixed an issue determining visibility when an element is hidden by an ancestor with a shared edge. Fixes [#27514](https://github.com/cypress-io/cypress/issues/27514).
```

**Review by chrisbreiding — CHANGES_REQUESTED** [review]

File: packages/driver/src/dom/visibility.ts [hunk 4]
```diff
@@ -552,8 +555,13 @@ export const getReasonIsHidden = function ($el, options = { checkOpacity: true }
 
   return `This element \`${node}\` is not visible.`
 }
```
Comment (emilyrohrbough) [inline packages/driver/src/dom/visibility.ts]:
```suggestion
}
/* eslint-enable no-cond-assign */
```

File: packages/driver/src/dom/visibility.ts (hunk 1 above)
Comment (emilyrohrbough) [reply packages/driver/src/dom/visibility.ts:555]: no

File: packages/driver/cypress/e2e/dom/visibility.cy.ts (hunk 2 above)
Comment (emilyrohrbough) [reply packages/driver/cypress/e2e/dom/visibility.cy.ts:822]: The name & test previously written was incorrect. The name update reflect other naming in this file,

The element being tested is:

```js
      this.$parentNoWidthHeightOverflowAuto = add(`\
<div style='width: 0; height: 0px; overflow: auto;'>
  <span>parent no size, overflow: auto</span>
</div>`)
```
is not actually visible. See screenshot:

Screen Shot 2023-10-26 at 1 15 42 PM

**Review by chrisbreiding — APPROVED** [review]

**Review by cacieprins — APPROVED** [review]

**Pull request merged**